contracts/src/v0.8/keystone/KeystoneForwarder.sol

```diff
@@ -15,19 +15,54 @@ contract KeystoneForwarder is IForwarder, ConfirmedOwner, TypeAndVersionInterfac
   /// @param data the data that was received
   error InvalidData(bytes data);
 
+  error RepeatedSigner();
+
+  error WrongNumberOfSignatures();
+
+  error InvalidDonId();
+  error InvalidSigner();
+
   uint256 private constant SELECTOR_LENGTH = 4;
   uint256 private constant REPORT_LENGTH = 64;
 
-  struct HotVars {
-    bool reentrancyGuard; // guard against reentrancy
-  }
+  bool internal s_reentrancyGuard; // guard against reentrancy
 
-  HotVars internal s_hotVars; // Mixture of config and state, commonly accessed
+  /// @notice Contains the signing address of each oracle
+  struct OracleSet {
+    bool exists;
+    uint8 f; // Number of faulty nodes allowed
+    address[] signers;
+    mapping(address => uint256) _positions; // 1-indexed to detect unset values
+  }
+  mapping(bytes4 donId => OracleSet) internal s_configs;
 
   mapping(bytes32 => address) internal s_reports;
 
   constructor() ConfirmedOwner(msg.sender) {}
 
+  uint256 internal constant MAX_ORACLES = 31;
+
+  // NOTE: we don't inherit OCR2Base since unlike aggregator we only care about signers, not transmitters
+  // and the signers don't fetch their config from the forwarder
+  function setConfig(bytes4 donId, uint8 f, address[] calldata signers) external nonReentrant {
```

Comment:
- I would check that signers.length >= 3*f+1
- I would check that f > 0
- I would use f for `exists` check.